zkevm-circuits/src/compress_circuit.rs

```diff
@@ -0,0 +1,498 @@
+
+mod decode {
+    use super::Part;
+    use eth_types::Field;
+    use gadgets::util::Expr;
+    use halo2_proofs::plonk::Expression;
+
+    // parts data => origin data
+    pub(crate) fn data_expr<F: Field>(parts: &[Part<F>]) -> Expression<F> {
+        parts.iter().fold(0.expr(), |acc, part| {
+            acc + part.data.clone() * part.decode_factor.clone()
+        })
+    }
+
+    // parts num_bits => origin num_bits
+    pub(crate) fn num_bits_expr<F: Field>(parts: &[Part<F>]) -> Expression<F> {
+        parts
+            .iter()
+            .fold(0.expr(), |acc, part| acc + part.num_bits.clone())
+    }
+}
+
+// input:
+//      A = 1011101(7)
+//      ~ = 1111111111101(13)
+// circuit:
+// ```
+// input | data     | num_bits | r_multi | shift_factor | decode_factor | data_rlc
+// ------|----------|--------- |---------|--------------|---------------|---------------------------------
+//  'A'  | 1011101  | 7        | 0       | 2^1          | 2^0           | 1011101
+//  'A'  |          | 0        | 0       | 0            | 0             | 1011101
+//  'A'  |          | 0        | 0       | 0            | 0             | 1011101
+//  'A'  |          | 0        | 0       | 0            | 0             | 1011101
+//  '~'  | 1        | 1        | 1       | 2^0          | 2^12          | 10111011
+//  '~'  | 11111111 | 8        | 1       | 2^0          | 2^4           | 10111011*r+11111111
+//  '~'  | 1101     | 4        | 0       | 2^4          | 2^0           | (10111011*r+11111111)*r+11010000
+//  '~'  |          | 0        | 0       | 0            | 0             | (10111011*r+11111111)*r+11010000
+// ```
+// N.B. we can use one row represent all 4 parts
+#[derive(Debug, Clone)]
+pub struct CompressCircuitConfig<F: Field> {
+    q_enable: Selector,
+    word_begin: Column<Advice>, // enable when word starts
+    input: Column<Advice>,      // input bitstream byte/row
+    // output parts:
+    // encoded value
+    data: Column<Advice>,
+    data_rlc: Column<Advice>,
+    // encoded bit length
+    num_bits: Column<Advice>,
+    // randomness multiplier selector
+    r_multi: Column<Advice>,
+    // the factor with which the part needs to be multiplied
+    shift_factor: Column<Advice>,
+    // the factor with which the part needs to be multiplied when
+    // we decode it
+    decode_factor: Column<Advice>,
+    // fixed huffman table
+    huffman_table: [TableColumn; 3], // [value, code, code_length]
+
+    _marker: PhantomData<F>,
+}
+
+/// Circuit configuration arguments
+pub struct CompressCircuitConfigArgs<F> {
+    // randomness used to do RLC
+    pub randomness: Expression<F>,
+}
+
+impl<F: Field> SubCircuitConfig<F> for CompressCircuitConfig<F> {
+    type ConfigArgs = CompressCircuitConfigArgs<F>;
+
+    fn new(
+        meta: &mut ConstraintSystem<F>,
+        CompressCircuitConfigArgs { randomness }: Self::ConfigArgs,
+    ) -> Self {
+        let q_enable = meta.selector();
+        let word_begin = meta.advice_column();
```

Comment:
I think this can be fixed, because it's just set to true every  `MAX_PARTS` rows. If it's fixed you don't need to do a boolean constraint on the value, if it's an advice column you'd have to add a boolean constraint and have additional constraints to only set it to true every `MAX_PARTS` rows, so a lot more to check.